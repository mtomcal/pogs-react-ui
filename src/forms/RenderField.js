import React from "react";

export const states = {
  PRISTINE: 'PRISTINE',
  TOUCHED: 'TOUCHED',
  INVALID: 'INVALID'
};

function inferState(touched, error) {
  if (touched) {
    if (error) {
      return states.INVALID;
    }
    return states.TOUCHED;
  }
  return states.PRISTINE;
}

export default React.createClass({
  getInitialState() {
    return {
      classNames: ''
    };
  },
  onFocus(event) {
    this.setState({classNames: ` is-focused`});
  },
  onDefocus(event) {
    this.setState({classNames: ``});
  },
  render() {
    const {input, label, type, meta: {touched, error}} = this.props;
    let errorMessageJsx = <span></span>;
    let invalidClass = '';

    switch (inferState(touched, error)) {
      case states.INVALID:
        invalidClass = ' has-error';
        errorMessageJsx = <span className="text-danger">{error}</span>;
        break;
      default:
        errorMessageJsx = <span></span>;
        break;
    }

    return (
      <div className={`form-group${invalidClass}${this.state.classNames}`} onFocus={this.onFocus} onBlur={this.onDefocus}>
        <label className="control-label" htmlFor={input.name}>{label}</label>
        <input {...input} className={`form-control`} type={type}/>
        {errorMessageJsx}
      </div>
    );
  },
})