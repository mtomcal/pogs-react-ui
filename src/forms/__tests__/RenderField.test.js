import React from 'react';
import RenderField from '../RenderField';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

describe('renderField', () => {
  test('renders default RenderField correctly', () => {
    const component = renderer.create(
      <RenderField label="Test" input={{name: 'test'}} meta={{touched: false}}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders error message if error is present', () => {
    const component = shallow(
      <RenderField label="Test" input={{name: 'test', value: 'afdadsfasdf'}} meta={{touched: true, error: 'should be something else'}}/>
    );
    const dataError = component.find('.text-danger').text();
    expect(dataError).toEqual('should be something else');
  });
});
