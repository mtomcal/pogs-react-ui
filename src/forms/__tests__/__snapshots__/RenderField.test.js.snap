exports[`renderField renders default RenderField correctly 1`] = `
<div
  className="form-group"
  onBlur={[Function]}
  onFocus={[Function]}>
  <label
    className="control-label"
    htmlFor="test">
    Test
  </label>
  <input
    className="form-control"
    name="test"
    type={undefined} />
  <span />
</div>
`;
